keps/sig-node/20191127-cri-containerstats-enhancements.md

```diff
@@ -0,0 +1,200 @@
+- [Design Details](#design-details)
+  - [Test Plan](#test-plan)
+  - [Graduation Criteria](#graduation-criteria)
+  - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+  - [Version Skew Strategy](#version-skew-strategy)
+- [Implementation History](#implementation-history)
+<!-- /toc -->
+
+## Summary
+
+This enhancement is to extend `CpuUsage` and `MemoryUsage` fields in
+`ContainerStats`. And also add new `ProcessUsage`/`NetworkUsage` field in
+`ContainerStats`. The Container Runtime Engine will cover more metrics about
+managed container and let the metric collectors focus on node-level metrics or
+monitoring metrics.
+
+## Motivation
+
+Since more and more sandbox container runtimes shows up, like `kata-container`/
+`gVisor`/ `Firecracker`, the way to collect metrics from sandbox container is
+quite different from runC. For the third-party metric collectors, they have to
+maintain the knowledge about how to collect metrics from each kind of container
+runtimes. And more, kubelet codebase focus on core-metrics export
+(defined in [Kubernetes monitoring architecture#Terminology](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/instrumentation/monitoring_architecture.md#terminology)),
+and there is no powerful standard interface for container level metrics.
+It will bring burden to end-users who wants to build powerful monitor pipeline
+(defined in [Kubernetes monitoring architecture#Executive Summary](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/instrumentation/monitoring_architecture.md#executive-summary)) for sandbox container runtimes.
+
+Based on this case, we can extend Container Runtime Interface(CRI) to define
+the container level metrics. Each container runtime engine, which implemented
+CRI, knows lifecycle and metrics about each managed container better than any
+external metric collector. For the third-party metric collectors, they just
+need to follow the `ContainerStats` API defined by CRI. With standard CRI,
+it is easy for the collector from monitor pipeline (defined in [Kubernetes monitoring architecture](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/instrumentation/monitoring_architecture.md#executive-summary)) to collect metrics from different kind of runtimes.
+
+### Goals
+
+- Enhance Container Runtime Interface(CRI) to provide more metrics about
+managed container
+
+### Non-Goals
+
+- Extend Core-Metrics (defined in [Kubernetes monitoring architecture#Terminology](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/instrumentation/monitoring_architecture.md#terminology)) for 
+
+## Proposal
+
+### Implementation Details/Notes/Constraints
+
+Extend `ContainerStats` message in CRI.
+
+`CpuUsage` will cover usage about kernel/user modes and per CPU.
+
+```protobuf
+// CpuUsage provides the CPU usage information.
+message CpuUsage {
```

Comment:
Will this message cover both cgroup v1 and v2?

Reply:
Tried to use common term to do summary. But there is still different between cgroup v1 and v2. For now, the changes follows cgroup v1 mostly.

Reply:
I think we are going to use separate structs for v2: https://github.com/containerd/containerd/pull/3799

You don't need to define v2 stats now, but probably better to document that separate structs shall be added for v2.

Reply:
cc @giuseppe @mrunalp